### ISSUE TITLE
Aggregator does not have an aggregated tensor for TensorKey(tensor_name='block_1.1.running_mean',                     origin='aggregator_plan.yaml_4ebcd576', round_number=55, report=False, tags=('model',))","grpc_status":2}"

### ISSUE THREAD

**Issue by CasellaJr** [intel/openfl#356]
During training, after several epochs (55 in this example, but this number is random) this errors happens.
```
ERROR    Collaborator failed with error: <_InactiveRpcError of RPC that terminated with:                                                  envoy.py:75
                            status = StatusCode.UNKNOWN
                            details = "Exception calling application: Aggregator does not have an aggregated tensor for
                    TensorKey(tensor_name='block_1.1.running_mean', origin='aggregator_plan.yaml_4ebcd576', round_number=55, report=False,
                    tags=('model',))"
                            debug_error_string = "{"created":"@1646124781.553286945","description":"Error received from peer
                    ipv4:172.20.8.151:59526","file":"src/core/lib/surface/call.cc","file_line":1062,"grpc_message":"Exception calling application:
                    Aggregator does not have an aggregated tensor for TensorKey(tensor_name='block_1.1.running_mean',
                    origin='aggregator_plan.yaml_4ebcd576', round_number=55, report=False, tags=('model',))","grpc_status":2}"
                    >:
                    Traceback (most recent call last):
                      File "/home/lmancuso/bruno/lib/python3.7/site-packages/openfl/component/envoy/envoy.py", line 73, in run
                        self._run_collaborator()
                      File "/home/lmancuso/bruno/lib/python3.7/site-packages/openfl/component/envoy/envoy.py", line 141, in _run_collaborator
                        col.run()
                      File "/home/lmancuso/bruno/lib/python3.7/site-packages/openfl/component/collaborator/collaborator.py", line 145, in run
                        self.do_task(task, round_number)
                      File "/home/lmancuso/bruno/lib/python3.7/site-packages/openfl/component/collaborator/collaborator.py", line 215, in do_task
                        required_tensorkeys
                      File "/home/lmancuso/bruno/lib/python3.7/site-packages/openfl/component/collaborator/collaborator.py", line 257, in
                    get_numpy_dict_for_tensorkeys
                        return {k.tensor_name: self.get_data_for_tensorkey(k) for k in tensor_keys}
                      File "/home/lmancuso/bruno/lib/python3.7/site-packages/openfl/component/collaborator/collaborator.py", line 257, in <dictcomp>
                        return {k.tensor_name: self.get_data_for_tensorkey(k) for k in tensor_keys}
                      File "/home/lmancuso/bruno/lib/python3.7/site-packages/openfl/component/collaborator/collaborator.py", line 331, in
                    get_data_for_tensorkey
                        require_lossless=True
                      File "/home/lmancuso/bruno/lib/python3.7/site-packages/openfl/component/collaborator/collaborator.py", line 367, in
                    get_aggregated_tensor_from_aggregator
                        self.collaborator_name, tensor_name, round_number, report, tags, require_lossless)
                      File "/home/lmancuso/bruno/lib/python3.7/site-packages/openfl/transport/grpc/client.py", line 84, in wrapper
                        response = func(self, *args, **kwargs)
                      File "/home/lmancuso/bruno/lib/python3.7/site-packages/openfl/transport/grpc/client.py", line 285, in get_aggregated_tensor
                        response = self.stub.GetAggregatedTensor(request)
                      File "/home/lmancuso/bruno/lib/python3.7/site-packages/grpc/_interceptor.py", line 221, in __call__
                        compression=compression)
                      File "/home/lmancuso/bruno/lib/python3.7/site-packages/grpc/_interceptor.py", line 257, in _with_call
                        return call.result(), call
                      File "/home/lmancuso/bruno/lib/python3.7/site-packages/grpc/_channel.py", line 333, in result
                        raise self
                      File "/home/lmancuso/bruno/lib/python3.7/site-packages/grpc/_interceptor.py", line 247, in continuation
                        compression=new_compression)
                      File "/home/lmancuso/bruno/lib/python3.7/site-packages/grpc/_channel.py", line 934, in with_call
                        return _end_unary_response_blocking(state, call, True, None)
                      File "/home/lmancuso/bruno/lib/python3.7/site-packages/grpc/_channel.py", line 826, in _end_unary_response_blocking
                        raise _InactiveRpcError(state)
                    grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:
                            status = StatusCode.UNKNOWN
                            details = "Exception calling application: Aggregator does not have an aggregated tensor for
                    TensorKey(tensor_name='block_1.1.running_mean', origin='aggregator_plan.yaml_4ebcd576', round_number=55, report=False,
                    tags=('model',))"
                            debug_error_string = "{"created":"@1646124781.553286945","description":"Error received from peer
                    ipv4:172.20.8.151:59526","file":"src/core/lib/surface/call.cc","file_line":1062,"grpc_message":"Exception calling application:
                    Aggregator does not have an aggregated tensor for TensorKey(tensor_name='block_1.1.running_mean',
                    origin='aggregator_plan.yaml_4ebcd576', round_number=55, report=False, tags=('model',))","grpc_status":2}"
```

I was doing image classification on MNIST using 1 director and 10 envoys with a non-iid data distribution.

**Comment by CasellaJr** [comment]
I have always this error after a random number of rounds. I used the DEBUG option. I note that each round I have this strange sentence: `DEBUG    Unable to get tensor from local store...attempting to retrieve from client  ` but it goes on until the crash. I mean each round in each envoy there are all these:
```
 DEBUG    Unable to get tensor from local store...attempting to retrieve from client                                               collaborator.py:291
           DEBUG    Requesting aggregated tensor TensorKey(tensor_name='__opt_state_0_45_istensor_exp_avg_sq',                               collaborator.py:365
                    origin='aggregator_plan.yaml_4ebcd576', round_number=2, report=False, tags=('model',))
           DEBUG    Disconnecting from gRPC server at node02:59526                                                                                 client.py:237
           DEBUG    Connecting to gRPC at node02:59526                                                                                             client.py:256
           DEBUG    Disconnecting from gRPC server at node02:59526                                                                                 client.py:237
           DEBUG    Attempting to retrieve tensor TensorKey(tensor_name='__opt_group_lengths', origin='aggregator_plan.yaml_4ebcd576',       collaborator.py:270
                    round_number=2, report=False, tags=('model',)) from local store
           DEBUG    Unable to get tensor from local store...attempting to retrieve from client                                               collaborator.py:291
           DEBUG    Requesting aggregated tensor TensorKey(tensor_name='__opt_group_lengths', origin='aggregator_plan.yaml_4ebcd576',        collaborator.py:365
                    round_number=2, report=False, tags=('model',))
           DEBUG    Disconnecting from gRPC server at node02:59526                                                                                 client.py:237
           DEBUG    Connecting to gRPC at node02:59526                                                                                             client.py:256
           DEBUG    Disconnecting from gRPC server at node02:59526                                                                                 client.py:237
           DEBUG    Attempting to retrieve tensor TensorKey(tensor_name='__opt_state_needed', origin='env_three', round_number=2,            collaborator.py:270
                    report=False, tags=('model',)) from local store
           DEBUG    Found tensor TensorKey(tensor_name='__opt_state_needed', origin='env_three', round_number=2, report=False,               collaborator.py:334
                    tags=('model',)) in local TensorDB
```
Until the error
```
DEBUG    Unable to get tensor from local store...attempting to retrieve from client                                               collaborator.py:291
           DEBUG    Requesting aggregated tensor TensorKey(tensor_name='block_1.1.weight', origin='aggregator_plan.yaml_4ebcd576',           collaborator.py:365
                    round_number=65, report=False, tags=('model',))
           DEBUG    Disconnecting from gRPC server at node02:59526                                                                                 client.py:237
           DEBUG    Connecting to gRPC at node02:59526                                                                                             client.py:256
           INFO     Response code: StatusCode.UNKNOWN                                                                                               client.py:59
           ERROR    Collaborator failed with error: <_InactiveRpcError of RPC that terminated with:                                                  envoy.py:75
                            status = StatusCode.UNKNOWN
                            details = "Exception calling application: Aggregator does not have an aggregated tensor for
                    TensorKey(tensor_name='block_1.1.weight', origin='aggregator_plan.yaml_4ebcd576', round_number=65, report=False,
                    tags=('model',))"
                            debug_error_string = "{"created":"@1646253414.129735296","description":"Error received from peer
                    ipv4:172.20.8.151:59526","file":"src/core/lib/surface/call.cc","file_line":1062,"grpc_message":"Exception calling application:
                    Aggregator does not have an aggregated tensor for TensorKey(tensor_name='block_1.1.weight',
                    origin='aggregator_plan.yaml_4ebcd576', round_number=65, report=False, tags=('model',))","grpc_status":2}"
                    >:
                    Traceback (most recent call last):
                      File "/home/lmancuso/bruno/lib/python3.7/site-packages/openfl/component/envoy/envoy.py", line 73, in run
                        self._run_collaborator()
                      File "/home/lmancuso/bruno/lib/python3.7/site-packages/openfl/component/envoy/envoy.py", line 141, in _run_collaborator
                        col.run()
                      File "/home/lmancuso/bruno/lib/python3.7/site-packages/openfl/component/collaborator/collaborator.py", line 145, in run
                        self.do_task(task, round_number)
                      File "/home/lmancuso/bruno/lib/python3.7/site-packages/openfl/component/collaborator/collaborator.py", line 215, in do_task
                        required_tensorkeys
                      File "/home/lmancuso/bruno/lib/python3.7/site-packages/openfl/component/collaborator/collaborator.py", line 257, in
                    get_numpy_dict_for_tensorkeys
                        return {k.tensor_name: self.get_data_for_tensorkey(k) for k in tensor_keys}
                      File "/home/lmancuso/bruno/lib/python3.7/site-packages/openfl/component/collaborator/collaborator.py", line 257, in <dictcomp>
                        return {k.tensor_name: self.get_data_for_tensorkey(k) for k in tensor_keys}
                      File "/home/lmancuso/bruno/lib/python3.7/site-packages/openfl/component/collaborator/collaborator.py", line 331, in
                    get_data_for_tensorkey
                        require_lossless=True
                      File "/home/lmancuso/bruno/lib/python3.7/site-packages/openfl/component/collaborator/collaborator.py", line 367, in
                    get_aggregated_tensor_from_aggregator
                        self.collaborator_name, tensor_name, round_number, report, tags, require_lossless)
                      File "/home/lmancuso/bruno/lib/python3.7/site-packages/openfl/transport/grpc/client.py", line 84, in wrapper
                        response = func(self, args, *kwargs)
                      File "/home/lmancuso/bruno/lib/python3.7/site-packages/openfl/transport/grpc/client.py", line 285, in get_aggregated_tensor
                        response = self.stub.GetAggregatedTensor(request)
                      File "/home/lmancuso/bruno/lib/python3.7/site-packages/grpc/_interceptor.py", line 221, in _call_
                        compression=compression)
                      File "/home/lmancuso/bruno/lib/python3.7/site-packages/grpc/_interceptor.py", line 257, in _with_call
                        return call.result(), call
                      File "/home/lmancuso/bruno/lib/python3.7/site-packages/grpc/_channel.py", line 333, in result
                        raise self
                      File "/home/lmancuso/bruno/lib/python3.7/site-packages/grpc/_interceptor.py", line 247, in continuation
                        compression=new_compression)
                      File "/home/lmancuso/bruno/lib/python3.7/site-packages/grpc/_channel.py", line 934, in with_call
                        return _end_unary_response_blocking(state, call, True, None)
                      File "/home/lmancuso/bruno/lib/python3.7/site-packages/grpc/_channel.py", line 826, in _end_unary_response_blocking
                        raise _InactiveRpcError(state)
                    grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:
                            status = StatusCode.UNKNOWN
                            details = "Exception calling application: Aggregator does not have an aggregated tensor for
                    TensorKey(tensor_name='block_1.1.weight', origin='aggregator_plan.yaml_4ebcd576', round_number=65, report=False,
                    tags=('model',))"
                            debug_error_string = "{"created":"@1646253414.129735296","description":"Error received from peer
                    ipv4:172.20.8.151:59526","file":"src/core/lib/surface/call.cc","file_line":1062,"grpc_message":"Exception calling application:
                    Aggregator does not have an aggregated tensor for TensorKey(tensor_name='block_1.1.weight',
                    origin='aggregator_plan.yaml_4ebcd576', round_number=65, report=False, tags=('model',))","grpc_status":2}"
                    >
           INFO     Send WaitExperiment request                                                                                            director_client.py:77
           INFO     WaitExperiment response has received                                                                                   director_client.py:79
[20:37:10] DEBUG    Sending health check status: name: "env_six"                                                                          director_client.py:126

[20:38:10] DEBUG    Sending health check status: name: "env_six"                                                                          director_client.py:126

[20:39:48] DEBUG    Sending health check status: name: "env_six"                                                                          director_client.py:126
```

**Comment by CasellaJr** [comment]
I attach the logs of an experiment with 10 envoys. The experiment crashes after 35 rounds with the usual error.
[LOGS.zip](https://github.com/intel/openfl/files/8190545/LOGS.zip)
Mad love, peace out

**Comment by CasellaJr** [comment]
I add the following details:
Using 2 envoys it easily reaches 100 rounds. However, increasing the number of envoys increases the probability of crash

**Comment by Giemp95** [comment]
Hello everybody,

I spent some time looking at the parallel execution flow of the OpenFL framework, and I might have spotted the cause of this unexpected behaviour.
In my opinion, the parallel aspect of the framework is not handled correctly. The main mechanism of task dispatching managed by the Aggregator does not seem to consider any synchronisation problems that may arise between Collaborators computing at different speeds. This can lead to the problem aforeahaed mentioned and other unexpected behaviours too. For example, is there a mechanism checking if every Collaborator involved in the training has sent their local tensor to the Aggregator at a certain point in time? According to what I have seen, there is no such control. This can result in a faster Collaborator sending its local tensor and asking for the aggregated one before any other Collaborator has got time to send its own, resulting in an erroneous behaviour from the framework. In my opinion, such a point of synchronisation should be implemented to assess that the retrieval tensor is actually the result of the aggregation of the tensors of all collaborators; this would indeed also resolve the issue reported in this discussion.
Another aspect that I would like to mention, to explicitly discuss how the framework could be improved, is the sleep time imposed by the Aggregator on the Collaborators on certain occasions. The presence of such an arbitrary sleep time is itself a sign of poor control over the parallel flow of the framework: by reducing or increasing it, the error rate in the communication changes notably, especially when the training involves many Collaborators. Such sleep should not exist in my opinion. Still, a synchronisation process should be built to avoid erroneous communications instead: the correct behaviour of a framework cannot be entrusted to an arbitrary waiting time assigned to the various parallel components taking part in the distributed task.
I will appreciate any consideration the developers and community have for this crucial topic.

Best regards,
Gianluca

**Comment by CasellaJr** [comment]
> Hello everybody,
> 
> I spent some time looking at the parallel execution flow of the OpenFL framework, and I might have spotted the cause of this unexpected behaviour. In my opinion, the parallel aspect of the framework is not handled correctly. The main mechanism of task dispatching managed by the Aggregator does not seem to consider any synchronisation problems that may arise between Collaborators computing at different speeds. This can lead to the problem aforeahaed mentioned and other unexpected behaviours too. For example, is there a mechanism checking if every Collaborator involved in the training has sent their local tensor to the Aggregator at a certain point in time? According to what I have seen, there is no such control. This can result in a faster Collaborator sending its local tensor and asking for the aggregated one before any other Collaborator has got time to send its own, resulting in the retrieval of the same tensor sent initially from the faster Collaborator. In my opinion, such a point of synchronisation should be implemented to assess that the retrieval tensor is actually the result of the aggregation of the tensors of all collaborators; this would indeed also resolve the issue reported in this discussion. Another aspect that I would like to mention, to explicitly discuss how the framework could be improved, is the sleep time imposed by the Aggregator on the Collaborators on certain occasions. The presence of such an arbitrary sleep time is itself a sign of poor control over the parallel flow of the framework: by reducing or increasing it, the error rate in the communication changes notably, especially when the training involves many Collaborators. Such sleep should not exist in my opinion. Still, a synchronisation process should be built to avoid erroneous communications instead: the correct behaviour of a framework cannot be entrusted to an arbitrary waiting time assigned to the various parallel components taking part in the distributed task. I will appreciate any consideration the developers and community have for this crucial topic.
> 
> Best regards, Gianluca

Hi everyone. I followed the advice of Gianluca. I increased the sleep in `openfl/component/aggregator/aggregator.py` line 234 from 10s to 20s and I reached 100 rounds. So, the problem is here.

**Comment by alexey-khorkin** [comment]
@CasellaJr, I think problem more complex. I used from 8 to 15 envoys and every time achieve 100+ rounds without any errors (Tensorflow_MNIST example, local federation setup). So could you please provide more info about federation setup? For example, whether you started the envoys from the same folder or from different ones? What about virtualenvs? Was it localhost setup or not? Thanks!

**Comment by CasellaJr** [comment]
My setup:
1. Run the PyTorchTinyImagetNet Tutorial present in the interactive_api inside openfl-tutorials
2. I use 10 envoys, each envoy runs on a different computer 
3. I installed in each envoy the virtualenv, following the instructions in the official documentation
4. So no, not in local, but 10 different computers

**Comment by aleksandr-mokrov** [comment]
@CasellaJr, could you try move  `self.round_number += 1` after `self._save_model(self.round_number, self.last_state_path)` and replace `self._save_model(self.round_number, self.last_state_path)` by `self._save_model(self.round_number+1, self.last_state_path)` here https://github.com/intel/openfl/blob/develop/openfl/component/aggregator/aggregator.py#L851?
The problem is that after we have incremented the round_number, the aggregator server starts to give tasks for new round to collaborators https://github.com/intel/openfl/blob/develop/openfl/component/aggregator/aggregator.py#L279 (that requests handles in different threads), and then collaborators asks aggregated model form the aggregator, but by this moment the aggregator may not have time to save the model. The more threads, the higher the likelihood that such an event will occur. Max available number of thread depends on the number of CPU (cpu_count), and the number of current working thread depends on the number of collaborator, more precisely, on the number of current requests from collaborators. 
We will fix it as soon as possible.

**Comment by CasellaJr** [comment]
Good morning Aleksandr. Thanks for the answer. I will try your advice asap.
Meanwhile, I am using this trick: I changed these lines https://github.com/intel/openfl/blob/develop/openfl/component/aggregator/aggregator.py#L368 in aggregator.py removing the ValueError. Basically, I wrote this:
```
while(nparray is None):
            self.logger.debug(f'Waiting for tensor_key')
            sleep(5)
            nparray = self.tensor_db.get_tensor_from_cache(agg_tensor_key)       
```
In this way I am achieving 100 rounds without problems. However, it is only a temp solution.
